ClassFieldTheory/IsNonarchimedeanLocalField/Unramified.lean

```diff
@@ -0,0 +1,320 @@
+      natDegree_map_eq_of_injective (FaithfulSMul.algebraMap_injective _ _),
+      ← Monic.natDegree_map (minpoly.monic h₁) (IsLocalRing.residue 𝒪[K]), h₃, key, h₈]
+  · conv_lhs => rw [← h₈]
+    exact minpoly.natDegree_le _
+
+variable {n} in
+@[simp] theorem finrank_unramifiedExtension (hn : n ≠ 0) :
+    Module.finrank K (UnramifiedExtension K n) = n :=
+  (finrank_unramifiedExtension_and_residue K hn).1
+
+variable {n} in
+@[simp] theorem f_unramifiedExtension (hn : n ≠ 0) :
+    f K (UnramifiedExtension K n) = n := by
+  refine le_antisymm ?_ (by simpa [hn, f] using (finrank_unramifiedExtension_and_residue K hn).2)
+  conv_rhs => rw [← finrank_unramifiedExtension K hn]
+  exact f_le_n _ _
+
+instance : IsUnramified K (UnramifiedExtension K n) := .mk <| by
+  obtain rfl | hn := eq_or_ne n 0
+  · exact e_eq_one_of_n_eq_one _ _ <| finrank_unramifiedExtension_zero K
+  rw [← Nat.mul_left_inj (ne_of_gt <| f_pos K (UnramifiedExtension K n)),
+    one_mul, e_mul_f_eq_n]
+  simp [hn]
+
+instance : HasEnoughRootsOfUnity K (Nat.card 𝓀[K] - 1) := by
+  obtain ⟨ζ, hζ⟩ := HasEnoughRootsOfUnity.exists_primitiveRoot 𝓀[K] (Nat.card 𝓀[K] - 1)
+  have := Finite.one_lt_card (α := 𝓀[K])
+  have : NeZero (Nat.card 𝓀[K] - 1) := .mk <| by grind
+  exact ⟨⟨_, hζ.map_of_injective teichmuller_injective⟩, inferInstance⟩
+
+variable (L : Type*) [Field L] [ValuativeRel L] [TopologicalSpace L] [IsNonarchimedeanLocalField L]
+  [Algebra K L] [ValuativeExtension K L]
+
+/-- If `Kn` denotes the unramified extension of `K` of degree `n`, then `Kn` embeds into `L` if
+`n ∣ f K L`. This is half of the universal property. -/
+theorem nonempty_unramifiedExtension_alghom_of_dvd_f (n : ℕ) (hn : n ∣ f K L) :
+    Nonempty (UnramifiedExtension K n →ₐ[K] L) := by
+  have hf0 := NeZero.mk (f_pos K L).ne'
+  have hn0 := NeZero.mk <| ne_zero_of_dvd_ne_zero hf0.out hn
+  have h₁ := Nat.pow_sub_one_dvd_pow_sub_one (Nat.card 𝓀[K]) hn
+  obtain ⟨ζ, hζ⟩ := HasEnoughRootsOfUnity.exists_primitiveRoot
+    (UnramifiedExtension K n) (Nat.card 𝓀[K] ^ n - 1)
+  have h₂ := pos_of_ne_zero <| card_residue_pow_sub_one_in_ne_zero K hn0.out
+  have h₃ := pos_of_ne_zero <| card_residue_pow_sub_one_in_ne_zero K hf0.out
+  refine IntermediateField.nonempty_algHom_of_adjoin_splits
+    (forall_eq.mpr ⟨.of_pow h₂ <| hζ.1 ▸ isIntegral_one,
+      .of_dvd (g := X ^ (Nat.card 𝓀[K] ^ n - 1) - C 1) ?_
+        (X_pow_sub_C_ne_zero h₂ _) ?_⟩)
+    (intermediateFieldTop_eq_adjoin_primitive_root K _ hζ).symm
+  · rw [f_spec'] at h₁ h₃
+    have := NeZero.mk h₃.ne'
+    have := HasEnoughRootsOfUnity.of_dvd L h₁
+    obtain ⟨ζ', hζ'⟩ := HasEnoughRootsOfUnity.exists_primitiveRoot L (Nat.card 𝓀[K] ^ n - 1)
+    simpa [Splits] using X_pow_sub_one_splits hζ'
+  · conv_rhs => exact show _ = map (algebraMap K L) (X ^ (Nat.card 𝓀[K] ^ n - 1) - C 1) by simp
+    rw [map_dvd_map', minpoly.dvd_iff]
+    simp [hζ.1]
+
+/-- The universal property of unramified extensions. -/
```

Comment:
It's not a universal property because there is no uniqueness ;-) Probably the actual universal property is something to do with lifting maps on residue fields. Maybe just write "universal property"? I'm not suggesting that you add some actual universal property, what you have is enough for what we need. I guess one could argue that a "more correct" unramified extension would not eat a natural n but a field extension of k[K] of degree n, but then of course this should then all be set up in some more generality (no need to assume residue field is finite, extension of residue fields should probably be some arbitrary separable algebraic extension or something etc etc).